pkg/flowaggregator/certificate.go

```diff
@@ -0,0 +1,180 @@
+		},
+		NotBefore:             validFrom,
+		NotAfter:              validFrom.Add(maxAge),
+		IsCA:                  true,
+		ExtKeyUsage:           []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth, x509.ExtKeyUsageServerAuth},
+		KeyUsage:              x509.KeyUsageDigitalSignature | x509.KeyUsageCertSign,
+		BasicConstraintsValid: true,
+	}
+	// generate private key for CA
+	caKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	// generate CA certificate
+	caCert, err := x509.CreateCertificate(rand.Reader, cert, cert, &caKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	caPEM := new(bytes.Buffer)
+	pem.Encode(caPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: caCert,
+	})
+
+	return cert, caKey, caPEM.Bytes(), err
+}
+
+func generateCertKey(caCert *x509.Certificate, caKey *rsa.PrivateKey, isServer bool, flowAggregatorDNSName string) ([]byte, []byte, error) {
+	var cert *x509.Certificate
+	if isServer {
+		cert = &x509.Certificate{
+			SerialNumber: big.NewInt(2),
+			Subject: pkix.Name{
+				CommonName: fmt.Sprintf("flow-aggregator-server-certificate@%d", time.Now().Unix()),
+			},
+			DNSNames:    []string{flowAggregatorDNSName},
+			NotBefore:   validFrom,
+			NotAfter:    validFrom.Add(maxAge),
+			ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+			KeyUsage:    x509.KeyUsageDigitalSignature,
+		}
+	} else {
+		cert = &x509.Certificate{
+			SerialNumber: big.NewInt(3),
+			Subject: pkix.Name{
+				CommonName: fmt.Sprintf("flow-aggregator-client-certificate@%d", time.Now().Unix()),
+			},
+			NotBefore:   validFrom,
+			NotAfter:    validFrom.Add(maxAge),
+			ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth},
+			KeyUsage:    x509.KeyUsageDigitalSignature,
+		}
+	}
+	// generate private key for certificate
+	certKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		return nil, nil, err
+	}
+	// sign the certificate using CA certificate and key
+	certBytes, err := x509.CreateCertificate(rand.Reader, cert, caCert, &certKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPEM := new(bytes.Buffer)
+	pem.Encode(certPEM, &pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: certBytes,
+	})
+
+	certKeyPEM := new(bytes.Buffer)
+	pem.Encode(certKeyPEM, &pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(certKey),
+	})
+
+	return certPEM.Bytes(), certKeyPEM.Bytes(), nil
+}
+
+func syncCAAndClientCert(caCert, clientCert, clientKey []byte, k8sClient kubernetes.Interface) error {
+	klog.Info("Syncing CA certificate with ConfigMap")
+	caConfigMap, err := k8sClient.CoreV1().ConfigMaps(CAConfigMapNamespace).Get(context.TODO(), CAConfigMapName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("error getting ConfigMap %s: %v", CAConfigMapName, err)
+	}
+	if caConfigMap.Data != nil && caConfigMap.Data[CAConfigMapKey] == string(caCert) {
+		return nil
```

Comment:
Does CA configmap in-sync mean ClientCert secret in-sync too?
I think for now the desired caCert will never equal to the current one as we always regenerate the cert before calling `syncCAAndClientCert`. However, since you add the check here, I guess you may want to retry this function in case of transient failure in future, then we shouldn't return immediately because secret may haven't be updated.

Reply:
Correct. Updated the info message. The reason for line 140-141 is for future when we support certificate rotation. Removed them for now. Thanks!